dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/LoopAnalysis.java

```diff
@@ -0,0 +1,202 @@
+        public boolean isLast() {
+            return this == containingLoop.getIterations().get(containingLoop.getIterations().size() - 1);
+        }
+
+        public List<Event> computeBody() {
+            final Event terminator = getIterationEnd().getSuccessor();
+
+            final List<Event> body = new ArrayList<>();
+            Event cur = getIterationStart();
+            do {
+                body.add(cur);
+            } while ((cur = cur.getSuccessor()) != terminator);
+            return body;
+        }
+    }
+
+    private static class LoopLabelInfo {
+        private String loopName;
+        private boolean isBound;
+        private int iterationIndex;
+    }
+
+    private final Map<Thread, ImmutableList<LoopInfo>> thread2LoopsMap = new HashMap<>();
+    public ImmutableList<LoopInfo> getLoopsOfThread(Thread thread) {
+        return thread2LoopsMap.get(thread);
+    }
+
+    private LoopAnalysis() {}
+
+
+    public static LoopAnalysis newInstance(Program program) {
+        final LoopAnalysis loopAnalysis = new LoopAnalysis();
+        loopAnalysis.run(program);
+        return loopAnalysis;
+    }
+
+
+    private void run(Program program) {
+        final Function<Thread, ImmutableList<LoopInfo>> loopFindingAlgo =
+                program.isUnrolled() ? this::findUnrolledLoopsInThread : this::findLoopsInThread;
+
+        program.getThreads().forEach(t -> this.thread2LoopsMap.put(t, loopFindingAlgo.apply(t)));
+    }
+
+    private ImmutableList<LoopInfo> findUnrolledLoopsInThread(Thread thread) {
+        final Map<String, LoopInfo> loopName2InfoMap = new HashMap<>();
+        final List<LoopInfo> loops = new ArrayList<>();
+
+        int loopCounter = 0;
+        for (Event e : thread.getEvents()) {
+            final LoopLabelInfo labelInfo = tryParseLoopLabel(e);
+            if (labelInfo == null) {
+                continue;
+            }
+
+            final String loopName = labelInfo.loopName;
+            if (labelInfo.iterationIndex == 1) {
+                // We start a new loop
+                final LoopInfo loopInfo = new LoopInfo();
+                loopInfo.loopNumber = loopCounter++;
+                loopInfo.loopName = loopName;
+                loopInfo.thread = thread;
+                loopInfo.isUnrolled = true;
+                // NOTE: Inner loops may get copied while unrolling the outer loop.
+                // In that case, the inner loops all may have identical label names
+                // So this put-operation may overwrite an old entry, which is okay!
+                loopName2InfoMap.put(loopName, loopInfo);
+                loops.add(loopInfo);
+            }
+
+            final LoopInfo loopInfo = loopName2InfoMap.get(loopName);
+            if (labelInfo.isBound) {
+                final LoopIterationInfo lastIter = loopInfo.loopIterationInfos.get(loopInfo.loopIterationInfos.size() - 1);
+                // We assume the bound marker to be directly before the loop bounding event.
```

Comment:
`LoopUnrolling` guarantees this now to mark the place where a loop bound is inserted. 
Again, if we had a notion of `EventAnnotation`, I would rather annotate the bound event itself to terminate the loop.
Note that currently it is not really possible to match a bound event with the loop it belongs to without using the marker event. But now that I think about it, it might be possible to copy the metadata of the loop header label into the bounding jump so that they both have the same `uId` which could be used for identification.